internal/pkg/config/config.go

```diff
@@ -284,27 +284,59 @@ func (fm funcMatcher) MatchFunction(path, receiver, name string) bool {
 	return fm.Package.MatchString(path) && fm.Receiver.MatchString(receiver) && fm.Method.MatchString(name)
 }
 
-var readFileOnce sync.Once
-var readConfigCached *Config
-var readConfigCachedErr error
+// readConfigResults reduces disk access across multiple ReadConfig calls.
+type readConfigResults struct {
+	once sync.Once
+	conf *Config
+	err  error
+}
 
-// ReadConfig reads, parses, and validates config file.
-func ReadConfig() (*Config, error) {
-	readFileOnce.Do(func() {
+// configCache safely stores readConfigResults for concurrent access.
+type configCache struct {
+	mu    sync.Mutex
+	cache map[string]*readConfigResults
+}
+
+// Instantiates readConfigResults if file has not yet been loaded
+func (w *configCache) read(file string) (*Config, error) {
+	w.instantiateCacheForFile(file)
+
+	cc := w.cache[file]
+	cc.once.Do(func() {
 		c := new(Config)
 		bytes, err := ioutil.ReadFile(configFile)
 		if err != nil {
-			readConfigCachedErr = fmt.Errorf("error reading analysis config: %v", err)
+			cc.err = fmt.Errorf("error reading analysis config: %v", err)
 			return
 		}
 
 		if err := yaml.UnmarshalStrict(bytes, c); err != nil {
-			readConfigCachedErr = err
+			cc.err = err
 			return
 		}
-		readConfigCached = c
+		cc.conf = c
 	})
-	return readConfigCached, readConfigCachedErr
+
+	return cc.conf, cc.err
+}
+
+func (w *configCache) instantiateCacheForFile(file string) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	if w.cache[file] == nil {
+		w.cache[file] = new(readConfigResults)
+	}
+}
+
+var cache = configCache{
+	mu:    sync.Mutex{},
```

Comment:
Nit: this can be omitted.

Reply:
I had thought the style was to specify all keys during initialization if you define any?  I can't seem to find where I read that, though, so I might be misremembering.

Reply:
I don't remember reading that in the style guides I've read. That feels like it could get annoying quickly if you have a struct with a lot of fields.

For context, there are places in the code where we don't respect this potential guideline, e.g. in `propagation.go:Taint`.

You may be thinking of the fact that if you key a field, the compiler won't let you have unkeyed fields in the same literal?

In any case. Explicitly setting defaults feels unnecessary to me, but this is not a big deal.

Reply:
Dunno!  Either way, zero-value assignment pruned.